xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+      new Thread.UncaughtExceptionHandler() {
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          logger.log(Level.SEVERE, "Exception!" + e);
+          // TODO(chengyuanzhang): implement cleanup.
+        }
+      });
+
+  private final String listenerAddress;
+  private final ServerBuilder<?> delegateBuilder;
+  private ScheduledExecutorService timeService = SharedResourceHolder.get(GrpcUtil.TIMER_SERVICE);
+  private final long retryDelayNano;
+  private final XdsClientPoolFactory xdsClientPoolFactory;
+  private final XdsServingStatusListener listener;
+  private final AtomicReference<FilterChainSelector> filterChainSelectorRef;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private final AtomicBoolean shutdown = new AtomicBoolean(false);
+  private boolean isServing;
+  private final CountDownLatch internalTerminationLatch = new CountDownLatch(1);
+  private final SettableFuture<IOException> initialStartFuture = SettableFuture.create();
+  private boolean initialStarted;
+  private ScheduledHandle restartTimer;
+  private ObjectPool<XdsClient> xdsClientPool;
+  private XdsClient xdsClient;
+  private DiscoveryState discoveryState;
+  private volatile Server delegate;
+
+  XdsServerWrapper(
+      String listenerAddress,
+      ServerBuilder<?> delegateBuilder,
+      long retryDelayNano,
+      XdsServingStatusListener listener,
+      AtomicReference<FilterChainSelector> filterChainSelectorRef,
+      XdsClientPoolFactory xdsClientPoolFactory) {
+    this.listenerAddress = checkNotNull(listenerAddress, "listenerAddress");
+    this.delegateBuilder = checkNotNull(delegateBuilder, "delegateBuilder");
+    this.retryDelayNano = retryDelayNano;
+    this.listener = checkNotNull(listener, "listener");
+    this.filterChainSelectorRef = checkNotNull(filterChainSelectorRef, "filterChainSelectorRef");
+    this.xdsClientPoolFactory = checkNotNull(xdsClientPoolFactory, "xdsClientPoolFactory");
+    this.delegate = delegateBuilder.build();
+  }
+
+  @Override
+  public Server start() throws IOException {
+    checkState(started.compareAndSet(false, true), "Already started");
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalStart();
+      }
+    });
+    syncContext.drain();
+    IOException exception;
+    try {
+      exception = initialStartFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (exception != null) {
+      throw exception;
+    }
+    return this;
+  }
+
+  private void internalStart() {
+    try {
+      xdsClientPool = xdsClientPoolFactory.getOrCreate();
+    } catch (Exception e) {
+      listener.onNotServing(
+          Status.UNAVAILABLE.withDescription(
+              "Failed to initialize xDS").withCause(e).asException());
+      initialStartFuture.set(null);
```

Comment:
If you set it to `null`, wouldn't `start()` proceed normally and return to the caller as if `start` was successful? Isn't there a test to catch this scenario? Also ` initialStartFuture` is defined as `SettableFuture<IOException>` : you can either define it as `SettableFuture<Exception>` or wrap all exceptions (such as this) as `IOException`s.

Reply:
@ejona86 @sanjaypujare 
When `xdsServerWrapper` return from initial `start()`, we may encounter various errors other than a real `IOException` that is thrown from `delegate.start()`(in this case we return but also keep retrying), e.g. creating xdsclient failure, bootstrap failure, `xdsClient.onResourceNotExist`, or `xdsClient.onError` that are permanent. 
In those cases, shall we wrap and throw an `IOException`  or just silently return? (We will by all means notify through `onNotServing`, and the interface only defines this exception type.)
The current behaviour was to wrap them with `IOException`.

Reply:
The doc for `Server.start()` says "After this call returns, clients may begin connecting to the listening socket(s)."  which won't happen in this case

Reply:
thats fair, let's wrap whatever exception as ioexception then, not changing current behavior.